### PR TITLE
Improve query result processing

### PR DESCRIPTION
**Pull request by pettyjamesm**
Improve performance of client / server query results handling by removing unnecessary copies allocations and overheads. Overall query execution time for a host-local trino client running a `SELECT * FROM tpch.sf100.lineitem` query to `/dev/null` and without compression enabled is reduced by 25%.

## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

Performance improvement achieved through refactoring

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Core engine and Client Libraries

> How would you describe this change to a non-technical end user or system administrator?

A non-technical end user or system administrator should not need to understand the details of this change.

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

### CODE REVIEW

File: client/trino-jdbc/src/main/java/io/trino/jdbc/Row.java [hunk 1]
```diff
@@ -34,9 +35,12 @@
 
     public List<RowField> getFields()
     {
-        return row.getFields().stream()
-                .map(RowField::new)
-                .collect(toImmutableList());
+        List<io.trino.client.RowField> fields = row.getFields();
```
Comment (lhofhansl) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/Row.java:38]: Did you benchmark this? Is this really faster?

File: client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java [hunk 2]
```diff
@@ -169,14 +169,14 @@
...
-                        Map<String, Object> result = new HashMap<>();
-                        for (RowField field : row.getFields()) {
-                            String name = field.getName()
-                                    .orElseGet(() -> "field" + field.getOrdinal());
-                            if (result.containsKey(name)) {
+                        List<RowField> fields = row.getFields();
+                        Map<String, Object> result = new HashMap<>(fields.size() + (fields.size() / 2));
+                        for (RowField field : fields) {
+                            Optional<String> fieldName = field.getName();
+                            String name = fieldName.isPresent() ? fieldName.get() : "field" + field.getOrdinal();
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java]: It's a capturing lambda that captures the loop variable, so an extra allocation on each iteration that could be avoided. I changed this slightly to in the latest version because of a bug uncovered by the unit tests.

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java [hunk 3]
```diff
@@ -63,15 +63,22 @@ private FixJsonDataUtils() {}
...
                 .map(Column::getTypeSignature)
-                .collect(toList());
-        ImmutableList.Builder<List<Object>> rows = ImmutableList.builder();
+                .toArray(ClientTypeSignature[]::new);
+        ImmutableList.Builder<List<Object>> rows;
+        if (data instanceof Collection<?>) {
+            rows = ImmutableList.builderWithExpectedSize(((Collection<?>) data).size());
+        }
+        else {
+            rows = ImmutableList.builder();
```
Comment (findepi) [inline client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:74]: data is always Collection, right? just change method signature.

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java [hunk 4]
```diff
@@ -63,15 +63,22 @@ private FixJsonDataUtils() {}
...
+        }
         for (List<Object> row : data) {
-            checkArgument(row.size() == columns.size(), "row/column size mismatch");
-            List<Object> newRow = new ArrayList<>();
-            for (int i = 0; i < row.size(); i++) {
-                newRow.add(fixValue(signatures.get(i), row.get(i)));
+            int rowColumns = row.size();
```
Comment (findepi) [inline client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:77]: Does extracting this variable actually matters? (it definitely impacts readability)

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java [hunk 5]
```diff
@@ -88,26 +95,29 @@ private static Object fixValue(ClientTypeSignature signature, Object value)
...
             }
             return fixedValue;
         }
         if (signature.getRawType().equals(MAP)) {
             ClientTypeSignature keySignature = signature.getArgumentsAsTypeSignatures().get(0);
             ClientTypeSignature valueSignature = signature.getArgumentsAsTypeSignatures().get(1);
-            Map<Object, Object> fixedValue = new HashMap<>();
-            for (Map.Entry<?, ?> entry : ((Map<?, ?>) value).entrySet()) {
+            Map<?, ?> mapValue = (Map<?, ?>) value;
+            Map<Object, Object> fixedValue = new HashMap<>(mapValue.size() + (mapValue.size() / 2));
```
Comment (findepi) [inline client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:109]: `Maps.newHashMapWithExpectedSize`

File: client/trino-client/src/main/java/io/trino/client/Row.java [hunk 6]
```diff
@@ -77,11 +76,25 @@ public static Builder builder()
         return new Builder();
     }
 
+    public static Builder builder(int fields)
```
Comment (findepi) [inline client/trino-client/src/main/java/io/trino/client/Row.java:79]: `builderWithExpectedSize`

File: client/trino-client/src/test/java/io/trino/client/TestFixJsonDataUtils.java [hunk 7]
```diff
@@ -79,7 +79,7 @@ public void testFixData()
                         field("foo", BIGINT),
                         field("bar", BIGINT))).getTypeSignature(),
                 ImmutableList.of(1, 2),
-                Row.builder()
+                Row.builder(2)
```
Comment (findepi) [inline client/trino-client/src/test/java/io/trino/client/TestFixJsonDataUtils.java:82]: revert

File: client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java [hunk 8]
```diff
@@ -169,10 +169,13 @@
                     .add("row", io.trino.client.Row.class, Row.class, (type, clientRow) -> (Row) convertFromClientRepresentation(type, clientRow))
                     .add("row", io.trino.client.Row.class, Map.class, (type, clientRow) -> {
                         Row row = (Row) convertFromClientRepresentation(type, clientRow);
-                        Map<String, Object> result = new HashMap<>();
-                        for (RowField field : row.getFields()) {
-                            String name = field.getName()
-                                    .orElseGet(() -> "field" + field.getOrdinal());
+                        List<RowField> fields = row.getFields();
+                        Map<String, Object> result = new HashMap<>(fields.size() + (fields.size() / 2));
```
Comment (findepi) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java:173]: `Maps.newHashMapWithExpectedSize`

File: client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java [hunk 9]
```diff
@@ -660,9 +663,10 @@ private static Object convertFromClientRepresentation(ClientTypeSignature column
                 io.trino.client.Row row = (io.trino.client.Row) value;
                 List<io.trino.client.RowField> fields = row.getFields();
                 List<ClientTypeSignatureParameter> typeArguments = columnType.getArguments();
-                Row.Builder builder = Row.builder();
-                verify(fields.size() == typeArguments.size(), "Type mismatch: %s, %s", row, columnType);
-                for (int i = 0; i < fields.size(); i++) {
+                int fieldCount = fields.size();
```
Comment (findepi) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java:666]: inline ?

File: client/trino-jdbc/src/main/java/io/trino/jdbc/Row.java [hunk 10]
```diff
@@ -69,12 +73,25 @@ public static Builder builder()
         return new Builder();
     }
 
+    public static Builder builder(int fields)
```
Comment (findepi) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/Row.java:76]: `builderWithExpectedSize`

File: core/trino-main/src/main/java/io/trino/server/protocol/QueryResultRows.java [hunk 11]
```diff
@@ -375,32 +375,34 @@ private Object getLegacyValue(Object value, Type type)
...
             }
 
             if (type instanceof MapType) {
                 Type keyType = ((MapType) type).getKeyType();
                 Type valueType = ((MapType) type).getValueType();
 
-                Map<Object, Object> result = new HashMap<>();
-                ((Map<Object, Object>) value).forEach((key, val) -> result.put(getLegacyValue(key, keyType), getLegacyValue(val, valueType)));
+                Map<Object, Object> mapValue = ((Map<Object, Object>) value);
+                Map<Object, Object> result = new HashMap<>(mapValue.size() + (mapValue.size() / 2));
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/server/protocol/QueryResultRows.java:392]: here too

File: core/trino-main/src/main/java/io/trino/server/protocol/QueryResultRows.java [hunk 12]
```diff
@@ -375,32 +375,34 @@ private Object getLegacyValue(Object value, Type type)
...
-                        .map(RowType.Field::getType)
-                        .collect(toImmutableList());
 
                 List<Object> result = new ArrayList<>(values.size());
                 for (int i = 0; i < values.size(); i++) {
-                    result.add(i, getLegacyValue(values.get(i), types.get(i)));
+                    result.add(getLegacyValue(values.get(i), fields.get(i).getType()));
                 }
-                return result;
+                return unmodifiableList(result);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/server/protocol/QueryResultRows.java:405]: unrelated?

**Review by findepi — COMMENTED** [review]
(just skimming)

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java (hunk 4 above)
Comment (pettyjamesm) [reply client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java]: Probably not, will leave it as `row.size()`

File: client/trino-client/src/main/java/io/trino/client/Row.java (hunk 6 above)
Comment (pettyjamesm) [reply client/trino-client/src/main/java/io/trino/client/Row.java]: Renamed

File: client/trino-client/src/test/java/io/trino/client/TestFixJsonDataUtils.java (hunk 7 above)
Comment (pettyjamesm) [reply client/trino-client/src/test/java/io/trino/client/TestFixJsonDataUtils.java]: Done

File: client/trino-jdbc/src/main/java/io/trino/jdbc/Row.java (hunk 10 above)
Comment (pettyjamesm) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/Row.java]: Renamed

File: client/trino-jdbc/src/main/java/io/trino/jdbc/Row.java (hunk 1 above)
Comment (wendigo) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/Row.java:38]: You can always use own collector definition with a builder having expected size:

```
    static <E> Collector<E, ?, ImmutableList<E>> toImmutableListWithExpectedSize(int expectedSize)
    {
        return (Collector) Collector.of(
                () -> ImmutableList.builderWithExpectedSize(expectedSize),
                ImmutableList.Builder::add,
                (builder, otherBuilder) -> builder.addAll(otherBuilder.build()),
                ImmutableList.Builder::build);
    }
```

**Comment by pettyjamesm** [comment]
@electrum - do you think you'll have a chance to review this soon?

**Comment by pettyjamesm** [comment]
@electrum, @martint - Any update? This PR was opened a month ago and has been sitting dormant for almost as long waiting for a review.

File: client/trino-client/src/main/java/io/trino/client/JsonResponse.java [hunk 13]
```diff
@@ -36,6 +38,7 @@
     private final int statusCode;
     private final String statusMessage;
     private final Headers headers;
+    @Nullable
```
Comment (arhimondr) [inline client/trino-client/src/main/java/io/trino/client/JsonResponse.java:41]: nit: We usually prefer `Optional<...>`. It causes an extra object allocation, but this is merely a single object per a single round trip of communication, shouldn't be a big deal.

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java [hunk 14]
```diff
@@ -51,27 +51,26 @@
...
 
-    public static Iterable<List<Object>> fixData(List<Column> columns, Iterable<List<Object>> data)
+    public static Iterable<List<Object>> fixData(List<Column> columns, List<List<Object>> data)
     {
         if (data == null) {
             return null;
         }
         requireNonNull(columns, "columns is null");
-        List<ClientTypeSignature> signatures = columns.stream()
+        ClientTypeSignature[] signatures = columns.stream()
```
Comment (arhimondr) [inline client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:65]: Why is this necessary?

**Review by arhimondr — APPROVED** [review]
LGTM % nits

File: client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java (hunk 14 above)
Comment (pettyjamesm) [reply client/trino-client/src/main/java/io/trino/client/FixJsonDataUtils.java:65]: It's not strictly speaking necessary, but we were already creating a new `List<ClientTypeSignature>` on each invokacation anyway, and the contents are accessed for each cell (each row, each column) in the output, so creating an array instead of a list is a minor efficiency improvement (it also avoids a check cast on each iteration incurred on the generic list implementation).

File: client/trino-client/src/main/java/io/trino/client/JsonResponse.java (hunk 13 above)
Comment (pettyjamesm) [reply client/trino-client/src/main/java/io/trino/client/JsonResponse.java:41]: True, I don't think it will matter much either way. I was aiming to keep consistency with the surrounding code where `exception` is nullable and `value` is nullable (either because the response was a `null` literal, or because there was an exception). I'm happy to change this to `Optional<String>` directly if you would prefer.

Comment (arhimondr) [reply client/trino-client/src/main/java/io/trino/client/JsonResponse.java:41]: Hmm. I'm fine with keeping it consistent with the surrounding code.
